### ISSUE TITLE
Some bugs on my frontend .. 

### ISSUE THREAD

**Issue by GuillaumeAV** [assemblee-virtuelle/mastopod#2]
Hey @srosset81 !
There are some problems on my screen :)
[Capture vidéo du 20-03-2024 14:38:36.webm](https://github.com/assemblee-virtuelle/mastopod/assets/6160028/aba364f2-54ed-4265-9e5d-6570379f0f07)

**Comment by srosset81** [comment]
Yes this is because you already have quite a few activities on your account which are not related to Mastopod.

It's not a big problem, but we need to improve the UX so that it is not visible to the user.

**Comment by SlyRock** [comment]
@srosset81 @Laurin-W 
Hi guys,
@Laurin-W it seems you proposed to use Shex validation as a filtering mechanism in order to fix this issue.

I have given some thoughts to this and it seems nice. 
I was thinking of implementing that on the `useCollection` hook, that could look like : 
```javascript
  // Define ShEx schema
  const shexSchema = `
    PREFIX as: <https://www.w3.org/ns/activitystreams#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

    start = @<ActivityShape>

    <ActivityShape> {
      rdf:type [as:Create as:Note as:Like as:Announce]+
    }
  `;

  const {
    items: activities,
    fetchNextPage,
    isFetchingNextPage,
    isLoading
  } = useCollection('outbox', { liveUpdates: true, dereferenceItems: true, shexSchema: shexSchema });
```

What do you think ?

**Comment by srosset81** [comment]
I find it's a very good idea to implement this feature in the `useCollection` hook, so that it may easily benefit other usages.

We need to take into account that in the coming months @mguihal is going to implement [collections filtering on the server side](https://github.com/assemblee-virtuelle/semapps/issues/1251), which will be much faster because the filtering will be done directly in the SPARQL query. So it would be better if the solution we choose takes this into account.

If we can use shapes for filtering (see the issue for details), then we'll probably need to send the URI of the shape in the header.
If we cannot... well, passing a shape to the useCollection is not going to be very useful.

**Comment by SlyRock** [comment]
>We need to take into account that in the coming months @mguihal is going to implement https://github.com/assemblee-virtuelle/semapps/issues/1251, which will be much faster because the filtering will be done directly in the SPARQL query. So it would be better if the solution we choose takes this into account.

So if we pass the shape to the useCollection hook, we can start by filtering on the frontend (short term) and then move on to relaying the shape to the backend, that's nice :) I started looking into the hooks code, it's quite complicated and I'm not sure I totally grasp the flow for now. But intuitively it seems feasible. As long as the signature of the useCollection is the same, the transition will be seamless when moving from client-side to server-side filtering.

> If we can use shapes for filtering (see the issue for details), then we'll probably need to send the URI of the shape in the header.
If we cannot... well, passing a shape to the useCollection is not going to be very useful.

I'm still a bit confused about shapes. I naively see them as expressions, so in my mind a shape is a string that can be stored in a file and accessed via an URI but can solso be dynamically written in some piece of code. Am I making sense here when thinking the shape could be passed as an expression (in a string) as much as a URI pointing to a resource containing the expression ?

```javascript
  // Case 1 : Define ShEx schema in a string
  const shexSchema = `
    PREFIX as: <https://www.w3.org/ns/activitystreams#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

    start = @<ActivityShape>

    <ActivityShape> {
      rdf:type [as:Create as:Note as:Like as:Announce]+
     //Insert logic here to account for needs specific to a particular case
    }
  `;

  const {
    items: activities,
    fetchNextPage,
    isFetchingNextPage,
    isLoading
  } = useCollection('outbox', { liveUpdates: true, dereferenceItems: true, shexSchema: shexSchema });

  // Case 2 : Use a static shape from an local or external repo
  const shexUri = 'https://myshaperepo/myShape';

  const {
    items: activities,
    fetchNextPage,
    isFetchingNextPage,
    isLoading
  } = useCollection('outbox', { liveUpdates: true, dereferenceItems: true, shexUri: shexUri });
```

**Comment by srosset81** [comment]
> Am I making sense here when thinking the shape could be passed as an expression (in a string) as much as a URI pointing to a resource containing the expression ?

For the useCollection hook, it would make no difference, but it would be easier to just pass the shape as an expression. For the backend however, if we decide to pass the shape in the headers, it would be a bad practice to try to pass the expression (plus the size of headers is limited), so we would need to rely on URIs. Note that the shape URI can be hosted directly in the app frontend, we don't need a shape repository (repositories are needed when several applications need to share exactly the same shape).

**Comment by srosset81** [comment]
Well, the shapes repository is still very much work in progress. But anyway, why rely on an external repository when it would work very well with an internal one ? If we need to create a different shape for every application (because applications will have different needs), it will be a lot of mess on that shape repository, that is already difficult to order.

**Comment by Laurin-W** [comment]
I see..

Maybe let's talk about it in person later :)
